soc/arm/atmel_sam/sam4s/soc.c

```diff
@@ -29,126 +30,63 @@
  */
 static ALWAYS_INLINE void clock_init(void)
 {
-	uint32_t reg_val;
+	/* Switch the main clock to the internal OSC with 12MHz */
+	soc_pmc_switch_mainck_to_fastrc(SOC_PMC_FAST_RC_FREQ_12MHZ);
 
-#ifdef CONFIG_SOC_ATMEL_SAM4S_EXT_SLCK
-	/* Switch slow clock to the external 32 KHz crystal oscillator. */
-	SUPC->SUPC_CR = SUPC_CR_KEY_PASSWD | SUPC_CR_XTALSEL_CRYSTAL_SEL;
+	/* Switch MCK (Master Clock) to the main clock */
+	soc_pmc_mck_set_source(SOC_PMC_MCK_SRC_MAIN_CLK);
 
-	/* Wait for oscillator to be stabilized. */
-	while (!(SUPC->SUPC_SR & SUPC_SR_OSCSEL)) {
-		;
-	}
-
-#endif /* CONFIG_SOC_ATMEL_SAM4S_EXT_SLCK */
+	EFC0->EEFC_FMR = EEFC_FMR_FWS(0);
+#if defined(ID_EFC1)
+	EFC1->EEFC_FMR = EEFC_FMR_FWS(0);
+#endif
 
-#ifdef CONFIG_SOC_ATMEL_SAM4S_EXT_MAINCK
-	/*
-	 * Setup main external crystal oscillator.
-	 */
+	soc_pmc_enable_clock_failure_detector();
 
-	/* Start the external crystal oscillator. */
-	PMC->CKGR_MOR = CKGR_MOR_KEY_PASSWD
-			/* Fast RC oscillator frequency is at 4 MHz. */
-			| CKGR_MOR_MOSCRCF_4_MHz
-			/*
-			 * We select maximum setup time. While start up time
-			 * could be shortened this optimization is not deemed
-			 * critical right now.
-			 */
-			| CKGR_MOR_MOSCXTST(0xFFu)
-			/* RC oscillator must stay on. */
-			| CKGR_MOR_MOSCRCEN
-			| CKGR_MOR_MOSCXTEN;
-
-	/* Wait for oscillator to be stabilized. */
-	while (!(PMC->PMC_SR & PMC_SR_MOSCXTS)) {
-		;
+	if (IS_ENABLED(CONFIG_SOC_ATMEL_SAM4S_EXT_SLCK)) {
+		soc_supc_slow_clock_select_crystal_osc();
 	}
 
-	/* Select the external crystal oscillator as the main clock source. */
-	PMC->CKGR_MOR = CKGR_MOR_KEY_PASSWD
-			| CKGR_MOR_MOSCRCF_4_MHz
-			| CKGR_MOR_MOSCRCEN
-			| CKGR_MOR_MOSCXTEN
-			| CKGR_MOR_MOSCXTST(0xFFu)
-			| CKGR_MOR_MOSCSEL;
-
-	/* Wait for external oscillator to be selected. */
-	while (!(PMC->PMC_SR & PMC_SR_MOSCSELS)) {
-		;
-	}
+	if (IS_ENABLED(CONFIG_SOC_ATMEL_SAM4S_EXT_MAINCK)) {
+		/*
+		 * Setup main external crystal oscillator.
+		 */
 
-	/* Turn off RC oscillator, not used any longer, to save power */
-	PMC->CKGR_MOR = CKGR_MOR_KEY_PASSWD
-			| CKGR_MOR_MOSCSEL
-			| CKGR_MOR_MOSCXTST(0xFFu)
-			| CKGR_MOR_MOSCXTEN;
+		/* We select maximum setup time.
+		 * While start up time could be shortened
+		 * this optimization is not deemed
+		 * critical now.
+		 */
+		soc_pmc_switch_mainck_to_xtal(false, 0xff);
 
-	/* Wait for the RC oscillator to be turned off. */
-	while (PMC->PMC_SR & PMC_SR_MOSCRCS) {
-		;
+		if (IS_ENABLED(CONFIG_SOC_ATMEL_SAM4S_WAIT_MODE)) {
+			/*
+			 * Instruct CPU to enter Wait mode instead of Sleep mode to
+			 * keep Processor Clock (HCLK) and thus be able to debug
+			 * CPU using JTAG.
+			 */
+			soc_pmc_enable_waitmode();
+		}
```

Comment:
_soc_pmc_enable_waitmode();_ is an independent configuration, right? Need review all SoC files.

Reply:
Every SoC had a similar `Kconfig` for this.

Reply:
Everything that we are doing is extract the duplicated code. If you already want to fix this you can create a generic entry at https://github.com/zephyrproject-rtos/zephyr/blob/915f0fd74c6065b39ab84b2ecafce57000c822a7/soc/arm/atmel_sam/Kconfig#L18

My concern is that the `soc_pmc_enable_waitmode();` may not even be part of clock initialization and could maybe be moved directly to the `z_arm_platform_init`. In your refactor somehow it depends on `CONFIG_SOC_ATMEL_<SOC>_EXT_MAINCK` now which not seems right.

Reply:
I did not change the logic for this, it already worked that way.

Reply:
Then it is a bug that became obvious now : ) Please, move out clock init to avoid future confusion. It is better if you configure before clock because allow debug attach faster.